internal/telemetry/tracker.go

```diff
@@ -96,8 +119,369 @@ func (t *Tracker) trackClusterDeploymentHeartbeat(ctx context.Context) error {
 		)
 		if err != nil {
 			errs = errors.Join(errs, fmt.Errorf("failed to track the heartbeat of the clusterDeployment %s/%s", clusterDeployment.Namespace, clusterDeployment.Name))
+
 			continue
 		}
 	}
 	return errs
 }
+
+// Helper function to recognize GPU resources
+func isGPUResource(resourceName corev1.ResourceName) bool {
+	return strings.Contains(string(resourceName), "gpu") ||
+		strings.Contains(string(resourceName), "nvidia.com") ||
+		strings.Contains(string(resourceName), "amd.com")
+}
+
+// Helper function to count non-system pods per node and return a map indexed by node name
+func countNonSystemPodsOnNodes(ctx context.Context, k8sClient client.Client) (map[string]map[string]int, error) {
+	// Get all pods in the cluster -- these will be filtered to just those assigned to the node
+	pods := &corev1.PodList{}
+	if err := k8sClient.List(ctx, pods); err != nil {
+		return nil, err
+	}
+
+	nodePodMetrics := make(map[string]map[string]int)
+
+	// Count pods
+	for _, pod := range pods.Items {
+		// Skip system namespace pods using rubric
+		if systemNamespaces[pod.Namespace] {
+			continue
+		}
+
+		// Skip pods not assigned to a node
+		if pod.Spec.NodeName == "" {
+			continue
+		}
+
+		// Initialize node metrics if not exists
+		if nodePodMetrics[pod.Spec.NodeName] == nil {
+			nodePodMetrics[pod.Spec.NodeName] = map[string]int{
+				"nodePodCount":           0,
+				"nodePodRunningCount":    0,
+				"nodePodGpuRequestCount": 0,
+				"nodePodGpuAllocCount":   0,
+				"containerCount":         0,
+			}
+		}
+
+		// Count total non-system pods and containers
+		nodePodMetrics[pod.Spec.NodeName]["nodePodCount"]++
+		nodePodMetrics[pod.Spec.NodeName]["containerCount"] += len(pod.Spec.Containers) + len(pod.Spec.InitContainers)
+
+		// Running pods
+		if pod.Status.Phase == corev1.PodRunning {
+			nodePodMetrics[pod.Spec.NodeName]["nodePodRunningCount"]++
+		}
+
+		// Pods requesting GPU -- need to check resources associated with Pod containeres
+		podRequestGPU := false
+		podAllocGPU := false
+		for _, container := range pod.Spec.Containers {
+			if container.Resources.Requests != nil {
+				for resourceName := range container.Resources.Requests {
+					if isGPUResource(resourceName) {
+						podRequestGPU = true
+						break
+					}
+				}
+			}
+			if container.Resources.Limits != nil {
+				for resourceName := range container.Resources.Limits {
+					if isGPUResource(resourceName) {
+						podAllocGPU = true
+						break
+					}
+				}
+			}
+		}
+
+		if podRequestGPU {
+			nodePodMetrics[pod.Spec.NodeName]["nodePodGpuRequestCount"]++
+		}
+		if podAllocGPU {
+			nodePodMetrics[pod.Spec.NodeName]["nodePodGpuAllocCount"]++
+		}
```

Comment:
same question about chances that there are gpu nodes and workloads requiring gpus on a mgmt cluster